libs/vkd3d/swapchain.c

```diff
@@ -2056,9 +2151,12 @@ static void dxgi_vk_swap_chain_present_iteration(struct dxgi_vk_swap_chain *chai
      * Non-FIFO swapchains will pump their frame latency handles through the fallback path of blit command being done.
      * Especially on Xwayland, the present ID is updated when images actually hit on-screen due to MAILBOX behavior.
      * This would unnecessarily stall our progress. */
-    if (chain->wait_thread.active && !chain->present.present_id_valid && chain->request.swap_interval > 0)
+    if (chain->wait_thread.active && !chain->present.present_id_valid &&
+        (chain->request.swap_interval > 0 || chain->present.low_latency_state.mode))
     {
-        chain->present.present_id += 1;
+        chain->present.present_id = (chain->present.low_latency_state.mode) ?
+            chain->request.frame_id : chain->present.present_id + 1;
```

Comment:
Like in https://github.com/doitsujin/dxvk/pull/3690#discussion_r1479812303, I think this needs a check that the number doesn't go backwards.

Also, it seems that the ID is only added for FIFO swapchains. Can you confirm that the markers functionality still work correctly when no present IDs are passed to QueuePresentKHR?  
(I have a custom layer implementation of ll2, and this made me realize that I was actually making the ill assumption that there will be a matching QueuePresent call for each frame ID that appeared in a marker.)

Reply:
For swap_interval > 1, it's possible to have multiple presents per DXGI present, so using the same frame_id over and over is also not allowed.

If LL2 is used, number of swap iterations needs to be forced to 1. I think that makes sense either way, since the frame latency mechanism will blow up when using the unrolled presentation.

Reply:
I added a path on the present thread to fallback to having vkd3d increment the present id if the application fails to provide a new frame id. If the application starts providing a frame id again the present task will recreate the swapchain and start using the application provided id again. I wouldn't expect apps to do this on purpose, but it is good to be robust against these types of issues. Thanks for the suggestion.

I think a present ID will be provided to vkQueuePresentKHR for FIFO swapchains, or if low latency mode is enabled with the change I have suggested. Is there something I am missing in regards to your question?

I have also limited the present task to doing a single present iteration when low latency mode is enabled.